src/main/java/org/reficio/p2/FeatureBuilder.java

```diff
@@ -0,0 +1,226 @@
+import javax.xml.parsers.ParserConfigurationException;
+
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang.StringUtils;
+import org.reficio.p2.bundler.ArtifactBundlerInstructions;
+import org.reficio.p2.utils.Utils;
+import org.reficio.p2.utils.XmlUtils;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NodeList;
+
+import com.google.common.collect.Multimap;
+
+import aQute.lib.osgi.Jar;
+
+public class FeatureBuilder {
+
+	public FeatureBuilder(P2FeatureDefinition p2FeatureDefintion, Multimap<P2Artifact, 
+			ArtifactBundlerInstructions>  bundlerInstructions, boolean generateSourceFeature, String timestamp) {
+		this.p2FeatureDefintion = p2FeatureDefintion;
+		this.bundlerInstructions = bundlerInstructions;
+		this.generateSourceFeature = generateSourceFeature;
+		this.featureTimeStamp = timestamp;
+	}
+
+	Multimap<P2Artifact, ArtifactBundlerInstructions>  bundlerInstructions;
+	P2FeatureDefinition p2FeatureDefintion;
+	boolean generateSourceFeature;
+	
+	public void generate(File destinationFolder) {
+		try {
+			File featureContent = new File(destinationFolder, this.getFeatureFullName());
+			featureContent.mkdir();
+			this.buildXml();
+			XmlUtils.writeXml(this.xmlDoc, new File(featureContent, "feature.xml"));
+			
+			//we must be generating the feature file from the pom
+			FileOutputStream fos = new FileOutputStream(new File(destinationFolder, this.getFeatureFullName()+".jar"));
+			Manifest mf = new Manifest();
+			JarOutputStream jar = new JarOutputStream(fos, mf);
+			addToJar(jar, featureContent);
+			
+		} catch (Exception e) {
+			throw new RuntimeException("Cannot generate feature", e);
+		}
+	}
+	
+	private void addToJar(JarOutputStream jar, File content) throws IOException
+	{
+		for (File f : FileUtils.listFiles(content, null, true) ) {
+			String fname = f.getPath().replace("\\", "/");
+			if (f.isDirectory()) {
+				if (!fname.endsWith("/")) {
+					fname = fname + "/";
+				}
+				JarEntry entry = new JarEntry(fname);
+				entry.setTime(f.lastModified());
+				jar.putNextEntry(entry);
+				jar.closeEntry();
+			} else {
+				//must be a file
+				JarEntry entry = new JarEntry(fname);
+				entry.setTime(f.lastModified());
+				jar.putNextEntry(entry);
+				jar.write( IOUtils.toByteArray(new FileInputStream(f)) );
+				jar.closeEntry();
+			}
+			
+
+		}
+	}
+	
+	//cache this so that the same timestamp is used
+	String featureTimeStamp;
+	
+	
+	String getQualifiedFeatureVersion() {
+		String v = this.p2FeatureDefintion.getVersion();
+		return v.replace("qualifier",featureTimeStamp);
+	}
+	
+	
+	String getFeatureFullName() {
+		String id = this.p2FeatureDefintion.getId();
+		if (generateSourceFeature) {
+			id = id + ".source";
+		}
+		return id + "_" + this.getQualifiedFeatureVersion();
+	}
+
+
+	Document xmlDoc;
+	void buildXml() throws ParserConfigurationException, FileNotFoundException {
+		xmlDoc = this.fetchOrCreateXml();
```

Comment:
Could we make this class stateless?
xmlDoc could be returned from this method.
It would be also great to break this class into smaller pieces